cdap-watchdog/src/test/java/co/cask/cdap/logging/write/LogCleanupTest.java

```diff
@@ -337,6 +351,118 @@ public void testRemoteLocationFactory() throws Exception {
     baseDir.lastModified();
   }
 
+  @Test
+  public void testCleanupRunWithoutMeta() throws Exception {
+    FileMetaDataManager fileMetaDataManager = injector.getInstance(FileMetaDataManager.class);
+    NamespacedLocationFactory namespacedLocationFactory = injector.getInstance(NamespacedLocationFactory.class);
+
+    // Deletion boundary
+    long deletionBoundary = System.currentTimeMillis() - RETENTION_DURATION_MS;
+    LOG.info("deletionBoundary = {}", deletionBoundary);
+
+    // Setup directories
+    LoggingContext dummyContext = new FlowletLoggingContext("ns", "app", "flw", "flwt", "run", "instance");
+
+    NamespaceMeta finalMetadata = new NamespaceMeta.Builder().setName("ns").build();
+    namespaceQueryAdmin.create(finalMetadata);
+
+    Location namespacedLogsDir = namespacedLocationFactory.get(Id.Namespace.from("ns")).append(logBaseDir);
+
+    Location contextDir = namespacedLogsDir.append("app").append("flw");
+    List<Location> withMetaFiles = Lists.newArrayList();
+    List<Location> noMetaFiles = Lists.newArrayList();
+    generateFiles(fileMetaDataManager, deletionBoundary, dummyContext, contextDir, withMetaFiles, noMetaFiles);
+
+    LogCleanup logCleanup = new LogCleanup(fileMetaDataManager, rootLocationFactory, namespaceQueryAdmin,
+                                           namespacedLocationFactory, logBaseDir, RETENTION_DURATION_MS, impersonator);
+    logCleanup.run();
+
+    for (Location location : withMetaFiles) {
+      Assert.assertFalse("Location " + location + " is not deleted!", location.exists());
+    }
+
+    for (Location location : noMetaFiles) {
+      Assert.assertFalse("Location " + location + " is not deleted!", location.exists());
+    }
+
+    // Assert all the files with/without metadata is deleted
+    for (int i = 0; i < 5; ++i) {
+      Location delDir = contextDir.append("2012-12-1" + i);
+      Assert.assertFalse("Location " + delDir + " is not deleted!", delDir.exists());
+    }
+
+    // Assert metadata for all deleted files is deleted
+    NavigableMap<Long, Location> remainingFilesMap = fileMetaDataManager.listFiles(dummyContext);
+    Set<Location> metadataForDeletedFiles =
+      Sets.intersection(new HashSet<>(remainingFilesMap.values()), new HashSet<>(withMetaFiles)).immutableCopy();
+    Assert.assertEquals(ImmutableSet.of(), metadataForDeletedFiles);
+  }
+
+  @Test
+  public void testCleanFilesWithoutMeta() throws Exception {
```

Comment:
I don't see this test being any different from test `testCleanupRunWithoutMeta()`. It would be good to test more use cases here, like 
* have files newer than `RETENTION_DURATION_MS` with and without meta
* reduce batch size for processing files on disk (to say 5), and make sure multiple runs are required to delete all files without meta
* do the same thing by reducing the scan size for log meta (to again 5)